File: src/main/java/no/hvl/tk/visual/debugger/server/endpoint/message/DebuggingWSMessage.java
package no.hvl.tk.visual.debugger.server.endpoint.message;

import static no.hvl.tk.visual.debugger.util.DiagramToJSONConverter.getMapper;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.intellij.openapi.diagnostic.Logger;

public class DebuggingWSMessage {

  private static final Logger LOGGER = Logger.getInstance(DebuggingWSMessage.class);

  private final DebuggingMessageType type;
  private final String content;

  private final String fileName;
  private final Integer line;

  public DebuggingWSMessage(final DebuggingMessageType type, final String content) {
    this(type, content, null, null);
  }

  public DebuggingWSMessage(
      DebuggingMessageType type, String content, String fileName, Integer line) {
    this.type = type;
    this.content = content;
    this.fileName = fileName;
    this.line = line;
  }

  // Getter needed for serialize
  public String getType() {
    return this.type.getTypeString();
  }

  public String getContent() {
    return this.content;
  }

  public String getFileName() {
    return fileName;
  }

  public Integer getLine() {
    return line;
  }

  public String serialize() {
    try {
      return getMapper().writeValueAsString(this);
    } catch (final JsonProcessingException e) {
      LOGGER.error(e);
      return "JsonProcessingException";
    }
  }
}

File: src/main/java/no/hvl/tk/visual/debugger/server/endpoint/message/DebuggingWSMessage.kt
package no.hvl.tk.visual.debugger.server.endpoint.message

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.intellij.openapi.diagnostic.Logger

class DebuggingWSMessage @JvmOverloads constructor(
    private val type: DebuggingMessageType,
    val content: String,
    val fileName: String? = null,
    val line: Int? = null
) {
    // Getter needed for serialize
    fun getType(): String {
        return type.typeString
    }

    fun serialize(): String {
        val mapper = ObjectMapper()
        mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL)
        try {
            return mapper.writeValueAsString(this)
        } catch (e: JsonProcessingException) {
            LOGGER.error(e)
            return "JsonProcessingException: $e"
        }
    }

    companion object {
        private val LOGGER = Logger.getInstance(
            DebuggingWSMessage::class.java
        )
    }
}
